File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.js
function solution(players, callings) {
    const hash = {};
    players.forEach((player, i) => hash[player] = i);
    // { mumu: 0, soe: 1, poe: 2, kai: 3, mine: 4 }
    
    callings.forEach((player) => {
        const prevRanking = hash[player]; // 3
        const passedPlayer = players[prevRanking - 1]; // 추월당한 선수
        
        // 배열 업데이트
        players[prevRanking] = passedPlayer;
        players[prevRanking - 1] = player;
        
        // hash 업데이트
        hash[passedPlayer] = prevRanking;
        hash[player] = prevRanking - 1;
    })
    return players;
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    positions = {name: i for i, name in enumerate(players)}
    # {'mumu': 0, 'soe': 1, 'poe': 2, 'kai': 3, 'mine': 4}
        
    for name in callings:
        idx = positions[name]
        prevName = players[idx - 1]
        
        # 딕셔너리 업데이트
        positions[name] = idx - 1
        positions[prevName] = idx
        
        # 배열 업데이트
        players[idx - 1], players[idx] = players[idx], players[idx - 1]
        
    return players